x/pool-incentives/keeper/distr_test.go

```diff
@@ -9,327 +9,367 @@ import (
-	mintCoins := sdk.Coins{mintCoin}
-	err := mintKeeper.MintCoins(suite.Ctx, mintCoins)
-	suite.NoError(err)
-
-	err = mintKeeper.DistributeMintedCoin(suite.Ctx, mintCoin) // this calls AllocateAsset via hook
-	suite.NoError(err)
-
-	distribution.BeginBlocker(suite.Ctx, abci.RequestBeginBlock{}, *suite.App.DistrKeeper)
-
-	feePool := suite.App.DistrKeeper.GetFeePool(suite.Ctx)
-	suite.Equal("40000stake", suite.App.BankKeeper.GetBalance(suite.Ctx, suite.App.AccountKeeper.GetModuleAddress(authtypes.FeeCollectorName), "stake").String())
-	suite.Equal(sdk.NewDecCoinsFromCoins(sdk.NewCoin("stake", sdk.NewInt(40000))).String(), feePool.CommunityPool.String())
-	suite.Equal("40000stake", suite.App.BankKeeper.GetBalance(suite.Ctx, suite.App.AccountKeeper.GetModuleAddress(distrtypes.ModuleName), "stake").String())
-
-	// Community pool should be increased
-	mintCoin = sdk.NewCoin("stake", sdk.NewInt(100000))
-	mintCoins = sdk.Coins{mintCoin}
-	err = mintKeeper.MintCoins(suite.Ctx, mintCoins)
-	suite.NoError(err)
-	err = mintKeeper.DistributeMintedCoin(suite.Ctx, mintCoin) // this calls AllocateAsset via hook
-	suite.NoError(err)
-
-	distribution.BeginBlocker(suite.Ctx, abci.RequestBeginBlock{}, *suite.App.DistrKeeper)
 
-	feePool = suite.App.DistrKeeper.GetFeePool(suite.Ctx)
-	suite.Equal("80000stake", suite.App.BankKeeper.GetBalance(suite.Ctx, suite.App.AccountKeeper.GetModuleAddress(authtypes.FeeCollectorName), "stake").String())
-	suite.Equal(feePool.CommunityPool.String(), sdk.NewDecCoinsFromCoins(sdk.NewCoin("stake", sdk.NewInt(80000))).String())
-	suite.Equal(sdk.NewCoin("stake", sdk.NewInt(80000)), suite.App.BankKeeper.GetBalance(suite.Ctx, suite.App.AccountKeeper.GetModuleAddress(distrtypes.ModuleName), "stake"))
+	for _, test := range tests {
+		suite.Run(test.name, func() {
+			suite.Setup()
+			keeper := suite.App.PoolIncentivesKeeper
+			mintKeeper := suite.App.MintKeeper
+			params := suite.App.MintKeeper.GetParams(suite.Ctx)
+			params.WeightedDeveloperRewardsReceivers = []minttypes.WeightedAddress{
+				{
+					Address: sdk.AccAddress([]byte("addr1---------------")).String(),
+					Weight:  sdk.NewDec(1),
+				},
+			}
+			suite.App.MintKeeper.SetParams(suite.Ctx, params)
+
+			suite.PrepareBalancerPool()
+
+			// LockableDurations should be 1, 3, 7 hours from the default genesis state.
+			lockableDurations := keeper.GetLockableDurations(suite.Ctx)
+			suite.Equal(3, len(lockableDurations))
+
+			for i, duration := range lockableDurations {
+				suite.Equal(duration, types.DefaultGenesisState().GetLockableDurations()[i])
+			}
+
+			feePoolOrigin := suite.App.DistrKeeper.GetFeePool(suite.Ctx)
+
+			// Create record
+			err := keeper.ReplaceDistrRecords(suite.Ctx, test.testingDistrRecord...)
+			suite.Require().NoError(err)
+
+			err = mintKeeper.MintCoins(suite.Ctx, sdk.NewCoins(test.mintedCoins))
+			suite.Require().NoError(err)
+
+			err = mintKeeper.DistributeMintedCoin(suite.Ctx, test.mintedCoins) // this calls AllocateAsset via hook
```

Comment:
What this test seems to be doing is testing `AllocateAsset` vicariously through mint's methods. As a result, this becomes more like a functional test.

I would like to suggest testing this without using mint keeper in general and by calling `AllocateAsset` directly

We can do so by sending some coins to the pool-incentives' module account balance in the test setup.

The fact that the balance of the pool-incentives keeper is updated upon calling `DistributeMintedCoin` is already tested in `x/mint` so there is no need to duplicate that here.

Please let me know what you think

Reply:
I just kept the old logic flow and put the test cases into a table. 
Thanks for letting me know that, it will make the test much easy to look.
Do we want to skip the hook part and call the AllocateAsset function directly?

Reply:
Sorry for the late reply. Yes, that's exactly what I meant. What we have right now looks great